src/main/kotlin/app/revanced/patcher/Patcher.kt

```diff
@@ -191,201 +67,138 @@ class Patcher(private val options: PatcherOptions) {
+                    throw PatchException("'$patchName' did not succeed previously")
 
-                logger.trace("Skipping '$patchName' because it has already been applied")
+                logger.trace("Skipping '$patchName' because it has already been executed")
 
-                return PatchResultSuccess()
+                return
             }
 
-            // recursively execute all dependency patches
+            // Recursively execute all dependency patches.
             patchClass.dependencies?.forEach { dependencyClass ->
                 val dependency = dependencyClass.java
 
-                val result = executePatch(dependency, executedPatches)
-                if (result.isSuccess()) return@forEach
-
-                return PatchResultError(
-                    "'$patchName' depends on '${dependency.patchName}' but the following error was raised: " +
-                            result.error()!!.let { it.cause?.stackTraceToString() ?: it.message }
-                )
+                try {
+                    executePatch(dependency, executedPatches)
+                } catch (throwable: Throwable) {
+                    throw PatchException(
+                        "'$patchName' depends on '${dependency.patchName}' " +
+                                "but the following exception was raised: ${throwable.cause?.stackTraceToString() ?: throwable.message}",
+                        throwable
+                    )
+                }
             }
 
             val isResourcePatch = ResourcePatch::class.java.isAssignableFrom(patchClass)
             val patchInstance = patchClass.getDeclaredConstructor().newInstance()
 
-            // TODO: implement this in a more polymorphic way
+            // TODO: implement this in a more polymorphic way.
             val patchContext = if (isResourcePatch) {
                 context.resourceContext
             } else {
-                context.bytecodeContext.also { context ->
-                    (patchInstance as BytecodePatch).fingerprints?.resolve(
-                        context,
-                        context.classes.classes
-                    )
+                context.bytecodeContext.apply {
+                    val bytecodePatch = patchInstance as BytecodePatch
+                    bytecodePatch.fingerprints?.resolve(this, classes)
                 }
             }
 
             logger.trace("Executing '$patchName' of type: ${if (isResourcePatch) "resource" else "bytecode"}")
 
-            return try {
-                patchInstance.execute(patchContext).also {
-                    executedPatches[patchName] = ExecutedPatch(patchInstance, it.isSuccess())
-                }
-            } catch (e: Exception) {
-                PatchResultError(e).also {
-                    executedPatches[patchName] = ExecutedPatch(patchInstance, false)
-                }
+            var success = false
+            try {
+                patchInstance.execute(patchContext)
```

Comment:
Currently, patches are executed without any way to stop or pause them. This can be solved using coroutines. The coroutines can be suspended or canceled entirely. This QoL change will allow pausing or stopping patching.

Reply:
Done.